### PR TITLE
Atari header improvements

### PR DESCRIPTION
**Pull request by billkendrick**

### CODE REVIEW

File: include/_antic.h [hunk 1]
```diff
@@ -33,50 +44,86 @@
...
 
-/* Define a structure with the antic register offsets */
+/* Define a structure with the ANTIC coprocessor's register offsets */
 struct __antic {
-    unsigned char   dmactl; /* direct memory access control */
-    unsigned char   chactl; /* character mode control */
+    unsigned char   dmactl; /* (W) direct memory access control */
+
+/* Playfield modes: */
+#define DMACTL_PLAYFIELD_NONE     0x00
```
Comment (groessler) [inline include/_antic.h]: Please move the `#define`s of possible register values out of and below the structure itself. If it's not clear by their name to which register they apply, add a comment line before.
Please do that for in all structures in all include files where applicable.

File: include/_antic.h [hunk 2]
```diff
@@ -86,13 +133,43 @@ struct __antic {
...
+
+/* General modifier */
+#define DL_DLI(x)       ((unsigned char)((x) | 128)) /* enable Display List Interrupt on this mode line; requires NMIEN set to enable DLIs */
+
 
-/* general modifier */
-#define DL_DLI(x)       ((unsigned char)((x) | 128))
+/* Macros for the beginning and end of functions used as Display List Interrupts */
+#define DLI_START asm("pha"); asm("txa"); asm("pha"); asm("tya"); asm("pha");
+#define DLI_END asm("pla"); asm("tay"); asm("pla"); asm("tax"); asm("pla"); asm("rti");
```
Comment (groessler) [inline include/_antic.h]: I've never used DLI interrupts from C. Could you explain their usage? (Maybe update the documentation.)

Comment (oliverschmidt) [reply include/_antic.h]: I'm having a hard time to believe that there's a place in cc65 for something beside https://cc65.github.io/doc/funcref.html#ss3.218 for that purpose.

File: include/_gtia.h [hunk 3]
```diff
@@ -61,14 +75,47 @@ struct __gtia_write {
...
-    unsigned char   vdelay; /* 0x1C: vertical delay */
+
+#define PRIOR_P03_PF03          0x01 /* Players 0-3, then Playfields 0-3, then background */
+#define PRIOR_P01_PF03_P23      0x02 /* Players 0-1, then Playfields 0-3, then Players 2-3, then background */
+#define PRIOR_PF03_P03          0x04 /* Playfields 0-3, then Players 0-3, then background */
+#define PRIOR_PF01_P03_PF23     0x08 /* Playfields 0-1, then Players 0-3, then Playfields 2-3, then background */
+
+#define PRIOR_5TH_PLAYER        0x10 /* Four missiles combine to be a 5th player (uses COLPF3) */
+#define PRIOR_OVERLAP_3RD_COLOR 0x20 /* Overlap of players 0 and 1, and of players 2 and 3, results in a third color
+                                        (else overlap is black). The resulting color is a logical OR of the two player colors. */
```
Comment (oliverschmidt) [inline include/_gtia.h]: Please follow the (very unusual) style for multi-line comments used in the cc65 code base.

File: include/_pia.h [hunk 4]
```diff
@@ -34,18 +39,45 @@
...
+
+
     unsigned char   portb;  /* port B data r/w */
+/* Paddle 4-7 triggers (per PORTB bits); only 400/800 had four controller ports */
+#define PORTB_PTRIG7 0x80
+#define PORTB_PTRIG6 0x40
+#define PORTB_PTRIG5 0x08
+#define PORTB_PTRIG4 0x04
+
+    /* See also: "JOY_xxx_MASK" in "atari.h" */
```
Comment (oliverschmidt) [inline include/_pia.h]: As far as I can see you didn't indent comment-only lines in the other files. So please don't do here too.

File: include/_pokey.h [hunk 5]
```diff
@@ -45,15 +53,71 @@ struct __pokey_write {
...
     unsigned char   audctl; /* audio control */
+
+#define AUDCTL_CLOCKBASE_15HZ 0x01 /* Switch main clock base from 64 KHz to 15 KHz */
+#define AUDCTL_HIGHPASS_CHAN2 0x02 /* Insert high pass filter into channel two, clocked by channel four */
+#define AUDCTL_HIGHPASS_CHAN1 0x04 /* Insert high pass filter into channel one, clocked by channel two */
+#define AUDCTL_JOIN_CHAN34    0x08 /* Join channels four and three (16 bit) */
+#define AUDCTL_JOIN_CHAN12    0x10 /* Join channels two and one (16 bit) */
+#define AUDCTL_CLOCK_CHAN3_179MHZ 0x20 /* Clock channel three with 1.79 MHz */
+#define AUDCTL_CLOCK_CHAN1_179MHZ 0x40 /* Clock channel one with 1.79 MHz */
+#define AUDCTL_9BIT_POLY      0x80 /* Makes the 17 bit poly counter into nine bit poly (see also: RANDOM) */
```
Comment (oliverschmidt) [inline include/_pokey.h]: Please align the values of the macros above.

File: include/atari.h [hunk 6]
```diff
@@ -93,7 +93,7 @@
 
 /* luminance values go from 0 (black) to 7 (white) */
 
-/* hue values */
+/* hue values (these can vary depending on TV standard (NTSC vs PAL), tint potentiometer settings, TV tint settings, emulator palette, etc. */
```
Comment (oliverschmidt) [inline include/atari.h]: Missing `)`.

File: include/_antic.h [hunk 7]
```diff
@@ -33,50 +44,86 @@
...
-    unsigned char   chactl; /* character mode control */
+    unsigned char   dmactl; /* (W) direct memory access control */
+
+/* Playfield modes: */
+#define DMACTL_PLAYFIELD_NONE     0x00
+#define DMACTL_PLAYFIELD_NARROW   0x01 /* e.g., 32 bytes per scanline with thick borders */
+#define DMACTL_PLAYFIELD_NORMAL   0x02 /* e.g., 40 bytes per scanline with normal borders */
+#define DMACTL_PLAYFIELD_WIDE     0x03 /* e.g., 48 bytes per scanline with no borders (overscan) */
+/* Other options: */
+#define DMACTL_DMA_MISSILES    0x04 /* if not set, GTIA's GRAFP0 thru GRAFP3 used for player shapes; if set, ANTIC's PMBASE will be used to fetch shapes via DMA */
```
[the remark this answers is not in the document]
Comment (greg-king5) [reply include/_antic.h]: Also, it might be better to put long comments on lines above a definition.

File: include/_antic.h (hunk 2 above)
Comment (billkendrick) [reply include/_antic.h]: I'll consider moving this to the function reference at some point, or perhaps just create some separate docs specific to cc65 for the Atari 8-bits (and hence also 5200).  I'm not an expert wrt this stuff, but I can probably gather some info from Action! language examples of DLIs, and the caveats.

Comment (oliverschmidt) [reply include/_antic.h]: @billkendrick: Have you read my comment above (https://github.com/cc65/cc65/pull/831#discussion_r247509274) ?

File: include/_antic.h [hunk 8]
```diff
@@ -33,50 +44,118 @@
...
+
+/* Register bits for NMIEN (enabling interrupts)
+** and NMIST (determining the cause for the NMI interrupt)
+*/
+
+#define NMIEN_DLI   0x80
+/* Display List Interrupts
+** Called on a modeline when "DL_DLI" bit is set the ANTIC instruction,
+** and jumps through VDSLST vector.
+*/
```
Comment (oliverschmidt) [inline include/_antic.h]: As far as I understand this comment refers to the define above. Why are comments sometimes below the "things" (like here) ? This both seem unusual and inconsistent in this pull request.

File: include/_pia.h [hunk 9]
```diff
@@ -4,9 +4,14 @@
 /*                                                                           */
 /*                Internal include file, do not use directly                 */
 /*                                                                           */
+/* The Peripheral Interface Adapter (PIA) chip provides parallel I/O         */
+/* interfacing; it was used in Atari 400/800 and Commodore PET family of     */
+/* computers, for joystick and interrupts.                                   */
+/* - Sources; various + Wikpedia article on "Peripheral Interface Adapter"   */
```
Comment (oliverschmidt) [inline include/_pia.h]: The `-` at the beginning of the line looks like an enumeration. In include/_antic.h you just put the "source" right behind the "text". I suggest to do the same here too.

File: include/_pokey.h [hunk 10]
```diff
@@ -47,13 +55,109 @@ struct __pokey_write {
...
+
+    unsigned char   potgo;  /* start paddle scan sequence (see "ALLPOT") */
     unsigned char   unuse1; /* unused */
     unsigned char   serout; /* serial port data output */
     unsigned char   irqen;  /* interrupt request enable */
     unsigned char   skctl;  /* serial port control */
 };
+
+
+
```
Comment (oliverschmidt) [inline include/_pokey.h]: We don't have three blank lines "inside" "sections".

File: include/_antic.h (hunk 8 above)
Comment (groessler) [reply include/_antic.h]: I also don't know if we need such detailed explanation here. If someone programs the hw directly I guess it can/should be assumed that he knows the hardware. So a simple comment like `/* enable Display List Interrupts */` should be sufficient.

**Comment by oliverschmidt** [comment]
From formal / style perspective this seems to look good :-) However, I don't know about the content...

File: include/_antic.h [hunk 11]
```diff
@@ -33,66 +44,206 @@
...
-    unsigned char   nmires; /* nmi reset/status */
+    unsigned char   chbase; /* (W) msb of character set base address */
+    unsigned char   wsync;  /* (W) wait for horizontal synchronization */
+    unsigned char   vcount; /* (R) vertical line counter */
+    unsigned char   penh;   /* (R) light pen horizontal position */
+    unsigned char   penv;   /* (R) light pen vertical position */
+    unsigned char   nmien;  /* (W) non-maskable interrupt enable */
+    unsigned char   nmires;
+    /* (W) ("NMIRES") nmi reset -- clears the interrupt request register; resets all of the NMI status together
+    ** (R) ("NMIST") nmi status -- holds cause for the NMI interrupt
```
Comment (groessler) [inline include/_antic.h]: You could use an anonymous union here to provide both `nmist` and `nmires`, like
```
union {
  unsigned char nmires;  /* (W) ... .*/
  unsigned char nmist;   /* (R) .... */
}
```

File: include/_pia.h [hunk 12]
```diff
@@ -43,9 +48,115 @@ struct __pia {
...
+#define PxCTL_BIT1             0x02
+
+/* (W) Controls PORTA addressing
+** One equals PORTA register; zero equals direction control register
+*/
+#define PxCTL_ADDRESSING       0x04
+
+#define PxCTL_BIT4             0x10 /* "Set to one" */
+#define PxCTL_BIT5             0x20 /* "Set to one" */
+#define PxCTL_BIT6             0x40 /* "Set to zero" */
```
Comment (groessler) [inline include/_pia.h]: These "Set to" comments appear a bit ambiguous to me. Do they mean these values are set by default (the system), or do I as programmer need to set them. Maybe rephrase to "Set to xxx by the system" or "Set to xxx by default".

File: include/atari.h [hunk 13]
```diff
@@ -86,14 +90,20 @@
...
+/* Luminance values go from 0 (black) to 7 (white) */
 
-/* hue values */
+/* Hue values */
+/* (These can vary depending on TV standard (NTSC vs PAL),
+** tint potentiometer settings, TV tint settings, emulator palette, etc.)
+*/
 #define HUE_GREY        0
 #define HUE_GOLD        1
 #define HUE_GOLDORANGE  2
```
Comment (groessler) [inline include/atari.h]: I guess these hue values could be moved to `_gtia.h` as you suggested in a recent email to cc65-devel.

File: include/_pia.h (hunk 12 above)
Comment (billkendrick) [reply include/_pia.h]: I admit I'm not totally clear. This is literally what they're labeled as in Mapping the Atari. I'll consult some other references to try & figure out precisely what they do.

File: include/_antic.h (hunk 11 above)
Comment (billkendrick) [reply include/_antic.h]: Nice! I admit I've not had any reason to use `union` since, like, college 25 years ago! :smile:

File: include/_pia.h [hunk 14]
```diff
@@ -43,9 +48,115 @@ struct __pia {
...
+#define PORTB_BASICROM         0x02
+
+/* If set, the corresponding LED is turned off. If clear, the LED will be on.
+** (1200XL only)
+*/
+#define PORTB_LED1             0x04
+#define PORTB_LED2             0x08
+
+
+/* On the XE series of computers, PORTB is a bank-selected memory control register (read/write): */
```
[the remark this answers is not in the document]
Comment (billkendrick) [reply include/_pia.h]: From a stock Atari perspective, the bits here are only used on the 130XE (not stock XLs or stock 65XE), as they related to extended memory (beyond 64KB).

There are other kinds of RAM upgrades, which I believe work differently. (e.g., I have a 256KB RAMBO XL on my 1200XL, and I believe it cannot do separate banking for ANTIC vs CPU, like the 130XE can.)

I admit I'm a bit inexperienced here. (I've personally owned XLs with 256KB for 20+ years, but in terms of utilizing them, it's only via existing software, or using a RAMdisk in MyDOS. Not directly programming for them (yet!))

[the remark this answers is not in the document]
Comment (billkendrick) [reply include/_pia.h]: Compute!'s "Mapping the Atari", appendix 12 "The XL/XE Memory Map", states:

"Bits 4-6 are reserved (unused) in the XL and 65XE. Bits 4 and 5 in the 130XE are used to enable bank switching."

File: include/_pia.h (hunk 12 above)
Comment (billkendrick) [reply include/_pia.h]: I gave it my best. I'm confident it's probably sufficient for most needs.

**Comment by groessler** [comment]
I've merged your changes.
Thanks for your contribution.

I see that PET also uses `_pia.h`. It would be nice if you could make a follow up PR to move the Atari specific bit definitions into `atari.h`. The 5200 doesn't have a PIA.

**Comment by groessler** [comment]
Could you please fix `_pia.h` and remove the Atari related stuff from it?

**Comment by groessler** [comment]
Bill, what's about "Could you please fix `_pia.h` and remove the Atari related stuff from it?"

regards,
chris

**Comment by billkendrick** [comment]
On Fri, Feb 15, 2019 at 11:05:59AM -0800, Christian Groessler wrote:
> Bill, what's about "Could you please fix `_pia.h` and remove the Atari related stuff from it?"

Hi Chris, et al, sorry about the delay!  We had some record-breaking
weather here last weekend, and lost power for 36 hours, so I didn't
get the chance.  I'll see if I can do so this evening, after the kids
go to bed.  (Real life trumps hobbies! ;) )

-bill!

**Comment by billkendrick** [comment]
> Bill, what's about "Could you please fix `_pia.h` and remove the Atari related stuff from it?"

Done in https://github.com/cc65/cc65/pull/847
